File: scripts/filesystem.ts
import { existsSync, lstatSync, mkdirSync, mkdtempSync, readFileSync, realpathSync, renameSync, rmSync, writeFileSync } from 'node:fs'
import { dirname, isAbsolute, join, relative, resolve, sep } from 'node:path'

function isOutside(root: string, candidate: string): boolean {
  const path = relative(root, candidate)
  return path === '..' || path.startsWith(`..${sep}`) || isAbsolute(path)
}

export function containedInput(recipePath: string, relativePath: string): string {
  if (isAbsolute(relativePath)) throw new Error('candidatesCsv must be relative to the adopted input file.')
  const root = realpathSync(dirname(recipePath))
  const lexical = resolve(root, relativePath)
  if (isOutside(root, lexical)) throw new Error('candidatesCsv must stay inside the adopted input directory.')
  if (lstatSync(lexical).isSymbolicLink()) throw new Error('candidatesCsv must not be a symbolic link.')
  const candidate = realpathSync(lexical)
  if (isOutside(root, candidate)) throw new Error('candidatesCsv must stay inside the adopted input directory.')
  return candidate
}

export function safeOutputDirectory(path: string): string {
  const absolute = resolve(path)
  if (lstatSync(absolute, { throwIfNoEntry: false })?.isSymbolicLink()) {
    throw new Error('Output directory must not be a symbolic link.')
  }
  mkdirSync(absolute, { recursive: true })
  return realpathSync(absolute)
}

export function writeManagedFiles(outputPath: string, files: Readonly<Record<string, string>>, force: boolean): void {
  const output = safeOutputDirectory(outputPath)
  for (const name of Object.keys(files)) {
    const target = join(output, name)
    if (lstatSync(target, { throwIfNoEntry: false })?.isSymbolicLink()) throw new Error(`Managed output must not be a symbolic link: ${name}`)
    if (existsSync(target) && !force) throw new Error('Output already exists; choose a new directory or explicitly pass --force.')
  }
  const staging = mkdtempSync(join(output, '.catalog-staging-'))
  const backup = mkdtempSync(join(output, '.catalog-backup-'))
  const installed: string[] = []
  const backedUp: string[] = []
  let preserveBackup = false
  try {
    for (const [name, content] of Object.entries(files)) writeFileSync(join(staging, name), content, { flag: 'wx' })
    for (const [name, content] of Object.entries(files)) {
      if (readFileSync(join(staging, name), 'utf8') !== content) throw new Error(`Failed to verify staged output: ${name}`)
    }
    try {
      for (const name of Object.keys(files)) {
        const target = join(output, name)
        if (existsSync(target)) {
          renameSync(target, join(backup, name))
          backedUp.push(name)
        }
      }
      for (const name of Object.keys(files)) {
        renameSync(join(staging, name), join(output, name))
        installed.push(name)
      }
    } catch (error) {
      for (const name of installed) rmSync(join(output, name), { force: true })
      try {
        for (const name of backedUp) renameSync(join(backup, name), join(output, name))
      } catch (restoreError) {
        preserveBackup = true
        throw new AggregateError([error, restoreError], `Failed to publish and restore catalog package; backup preserved at ${backup}`)
      }
      throw error
    }
  } finally {
    rmSync(staging, { recursive: true, force: true })
    if (!preserveBackup) rmSync(backup, { recursive: true, force: true })
  }
}

File: scripts/catalog_sources/filesystem.py
"""Filesystem guards for catalog authoring outputs."""

import os
import shutil
import tempfile
from contextlib import contextmanager
from pathlib import Path
from typing import BinaryIO, Iterator


def safe_output_directory(path: Path) -> Path:
    absolute = path.absolute()
    if absolute.is_symlink():
        raise ValueError(f"Output directory must not be a symbolic link: {path}")
    absolute.mkdir(parents=True, exist_ok=True)
    return absolute.resolve(strict=True)


def managed_path(directory: Path, name: str) -> Path:
    root = safe_output_directory(directory)
    target = root / name
    if target.is_symlink():
        raise ValueError(f"Managed output must not be a symbolic link: {name}")
    return target


@contextmanager
def atomic_binary_writer(path: Path) -> Iterator[BinaryIO]:
    target = managed_path(path.parent, path.name)
    descriptor, temporary_name = tempfile.mkstemp(prefix=f".{target.name}.", dir=target.parent)
    temporary = Path(temporary_name)
    try:
        with os.fdopen(descriptor, "wb") as output:
            yield output
            output.flush()
            os.fsync(output.fileno())
        os.replace(temporary, target)
    finally:
        temporary.unlink(missing_ok=True)


def atomic_write_text(path: Path, content: str) -> None:
    with atomic_binary_writer(path) as output:
        output.write(content.encode("utf-8"))


def write_managed_files(directory: Path, files: dict[str, bytes | str], force: bool = False) -> None:
    root = safe_output_directory(directory)
    targets = {name: managed_path(root, name) for name in files}
    if not force and any(target.exists() for target in targets.values()):
        raise FileExistsError("Output exists; use --force for an intentional regeneration")
    staging = Path(tempfile.mkdtemp(prefix=".catalog-staging-", dir=root))
    backup = Path(tempfile.mkdtemp(prefix=".catalog-backup-", dir=root))
    installed: list[str] = []
    backed_up: list[str] = []
    preserve_backup = False
    try:
        for name, content in files.items():
            data = content.encode("utf-8") if isinstance(content, str) else content
            target = staging / name
            with target.open("xb") as output:
                output.write(data)
                output.flush()
                os.fsync(output.fileno())
            if target.read_bytes() != data:
                raise OSError(f"Failed to verify staged output: {name}")
        try:
            for name, target in targets.items():
                if target.exists():
                    os.replace(target, backup / name)
                    backed_up.append(name)
            for name, target in targets.items():
                os.replace(staging / name, target)
                installed.append(name)
        except BaseException as publish_error:
            for name in installed:
                targets[name].unlink(missing_ok=True)
            try:
                for name in backed_up:
                    os.replace(backup / name, targets[name])
            except BaseException as restore_error:
                preserve_backup = True
                raise ExceptionGroup(
                    f"Failed to publish and restore catalog package; backup preserved at {backup}",
                    [publish_error, restore_error],
                )
            raise
    finally:
        shutil.rmtree(staging, ignore_errors=True)
        if not preserve_backup:
            shutil.rmtree(backup, ignore_errors=True)